src/test/java/com/aws/greengrass/builtin/services/pubsub/SubscriptionTrieTest.java

```diff
@@ -0,0 +1,126 @@
+            assertThat(trie.get(topic), containsInAnyOrder(cb1, cb2));
+        }
+    }
+
+    static Stream<Arguments> subscriptionMatch() {
+        return Stream.of(
+                arguments("foo", singletonList("foo")),
+                arguments("foo/bar", singletonList("foo/bar")),
+                // multilevel wildcard #
+                arguments("#", asList("foo", "foo/bar", "foo/bar/baz", "$foo/bar")),
+                arguments("foo/#", asList("foo/bar", "foo/bar/baz", "foo/bar/#")),
+                // single level wildcard +
+                arguments("+", asList("foo", "foo/", "$foo")),
+                arguments("foo/+/baz", asList("foo/bar/baz")),
+                arguments("foo/+/baz/#", asList("foo//baz/bar", "foo/bar/baz/bat", "foo/bar/baz/bat/#"))
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("subscriptionNotMatch")
+    public void GIVEN_subscription_THEN_do_not_match(String subscription, List<String> topics) {
+        Object cb1 = new Object();
+        Object cb2 = new Object();
+        trie.add(subscription, cb1);
+        trie.add(subscription, cb2);
+
+        for (String topic : topics) {
+            assertThat(trie.get(topic), not(containsInAnyOrder(cb1, cb2)));
+        }
+    }
+
+    static Stream<Arguments> subscriptionNotMatch() {
+        return Stream.of(
+                arguments("foo", asList("fo", "foo/bar", "abc")),
+                arguments("foo/bar", asList("foo", "foo/bar/baz", "fo/bar")),
+                // multilevel wildcard # does not match parent topic
+                arguments("foo/#", asList("foo", "foo/")),
+                // single level wildcard +
+                arguments("+", asList("/foo", "foo/bar")),
+                arguments("foo/+/baz", asList("foo", "foo/baz", "foo/bar/bat/baz"))
+        );
+    }
+
+    @Test
+    public void GIVEN_subscription_WHEN_remove_topic_THEN_no_matches() {
+        Object cb1 = new Object();
+        Object cb2 = new Object();
+        String topic = "foo";
+        trie.add(topic, cb1);
+        trie.add(topic, cb2);
+        assertTrue(trie.containsKey(topic));
+        assertThat(trie.get(topic), containsInAnyOrder(cb1, cb2));
+        assertEquals(trie.size(), 2);
+
+        assertThat("remove topic", trie.remove(topic, cb1), is(true));
+        assertThat(trie.get(topic), contains(cb2));
+        assertEquals(trie.size(), 1);
+        assertThat("remove topic", trie.remove(topic, cb2), is(true));
+        assertThat(trie.get(topic), is(empty()));
+        assertEquals(trie.size(), 0);
+    }
+
+    @Test
+    public void GIVEN_subscription_wildcard_WHEN_remove_topic_THEN_no_matches() {
+        Object cb1 = new Object();
+        Object cb2 = new Object();
+        String topic = "foo/+/bar/#";
+        trie.add(topic, cb1);
+        trie.add(topic, cb2);
+        assertTrue(trie.containsKey(topic));
+        assertThat(trie.get(topic), containsInAnyOrder(cb1, cb2));
+        assertEquals(trie.size(), 2);
```

Comment:
These assertions are backwards. Order is expected and then the actual value.

Reply:
thanks